File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int dfs(int i, vector<int> numbers, int end, int sum, int target);

int solution(vector<int> numbers, int target) {
    int answer = 0;
    
    answer = dfs(0, numbers, numbers.size(), 0, target);
    
    return answer;
}

int dfs(int i, vector<int> numbers, int end, int sum, int target){
    
    if(i >= end){
        if(sum == target) return 1;
        else return 0;
    }
    else{
        return dfs(i+1, numbers, end, sum + numbers[i], target) + dfs(i+1, numbers, end, sum - numbers[i], target);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(int[] numbers, int target) {
        int count = 0;
        Queue<(int index, int sum)> queue = new Queue<(int, int)>();
        
        queue.Enqueue((0, 0));
        
        while(queue.Count > 0){
            var (idx, sum) = queue.Dequeue();
            
            if(idx == numbers.Length){
                if(sum == target) count++;
                continue;
            }
            
            int num = numbers[idx++];
            
            queue.Enqueue((idx, sum + num));
            queue.Enqueue((idx, sum - num));
        }
        
        return count;
    }
}
